Lambda Provisioned Concurrency (#2167)
pulumi/docs: content/blog/provisioned-concurrency-insurance-against-cold-starts-in-aws-lambda/index.md

---
title: "Provisioned Concurrency: Insurance Against Cold Starts in AWS Lambda"
date: 2019-12-18
meta_desc: "Using Pulumi to configure AWS Lambda provisioned concurrency"
meta_image: meta.png
authors: ["mikhail-shilkov"]
tags: ["AWS", "Serverless"]
---

> AWS has recently [announced](https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/) the launch of **Provisioned Concurrency**, a new feature of AWS Lambda that intends to solve the problem of cold starts. In this article, we take another look at the problem of latency-critical serverless applications, and how the new feature impacts the status-quo.

Reviewer: ```suggestion
AWS recently [announced](https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/) the launch of **Provisioned Concurrency**, a new feature of AWS Lambda that intends to solve the problem of cold starts. In this article, we take another look at the problem of latency-critical serverless applications, and how the new feature impacts the status-quo.
```

<!--more-->

## Concurrency Model of AWS Lambda

Despite being serverless, AWS Lambda uses lightweight containers to process incoming requests. Every container, or worker, can process only a single request at any given time.

{{< figure src="./executions.png" caption="Overlapping executions land on separate workers" >}}

> Therefore, the number of concurrent requests define the number of required workers that a specific AWS Lambda needs to serve at any given moment.

Reviewer: ```suggestion
Because of this, the number of concurrent requests defines the number of required workers that a specific AWS Lambda function needs to serve a response at any given moment.
```

Reviewer 2: @spara Are resolved suggestions applied or discarded?

Author: They should be applied, fixing these now.

## Cold Starts

How does AWS know how many workers it needs to run for a given function? Well, it doesn't know in advance. AWS allocates new workers on-demand as the Lambda gets invoked.

> Whenever Lambda receives a request but it has no idle workers, the control plane assigns a new generic worker to it. The worker then has to download the custom code or binaries of your Lambda and load them into memory before it can do service the request. This process takes time, which significantly increases response latency.

Reviewer: ```suggestion
Whenever Lambda receives a request but has no idle workers, the control plane assigns a new generic worker to it. The worker then has to download the custom code or binaries of your Lambda and load them into memory before it can do service the request. This process takes time, which significantly increases response latency.
```

> The issue of sporadically slow responses caused by the need to increase the pool of workers is known as **Cold Start**. Cold starts are consistently the top concern about the applicability of serverless tech to latency-sensitive workloads. There are numerous articles about the problem, including many articles I have written in the [Cold Starts](https://mikhail.io/serverless/coldstarts/) section on my website

Reviewer: ```suggestion
The issue of sporadically slow responses caused by the need to increase the pool of workers is known as **Cold Start**. Cold starts are consistently the top concern about the applicability of serverless technologies to latency-sensitive workloads. There are numerous articles about the problem, including many articles I have written in the [Cold Starts](https://mikhail.io/serverless/coldstarts/) section on my website
```

A trick known as **Lambda Warmers** uses kept-alive workers to reduce the frequency of cold starts. The idea is to have a CloudWatch timer that fires every few minutes and sends `N` parallel requests to the target Lambda. If all those requests land at the same time, AWS has to provision at least `N` workers to process them. The actual operation doesn't have to do any useful work, but it resets the recycling timer back to zero.

There are a few drawbacks with this approach:

* If a valid request comes at the same time as warming requests, it might hit a cold start.
* Extra logic is needed in the Lambda code to detect warming requests and short-circuit instead of doing useful work.
* CloudWatch rules require additional setup and management.
* The result is still best-effort: Lambda still recycles workers from time to time.

You can find more details about how Pulumi can help with some of these issues in [AWS Lambda Warmer as Pulumi Component](https://mikhail.io/2018/08/aws-lambda-warmer-as-pulumi-component/).

> Here is an example of configuring the provisioned concurrency with Pulumi in TypeScript:

Reviewer: This is a bit of an abrupt transition - shouldn't there be a section header with an introduction to what Provisioned Concurrency is, how it works, etc? and then the code example?

Reviewer 2: Wow, definitely not intended. A title and a paragraph from my original doc are gone. Restoring!

Author: I don't see the transition in the google doc across versions.

```ts
const lambda = new aws.lambda.Function("mylambda", {
    code: new pulumi.asset.FileArchive("./app"),
    handler: "handler",
    role: role.arn,
    publish: true,
});

const fixed = new aws.lambda.ProvisionedConcurrencyConfig("fixed-concurrency", {
  functionName: lambda.name,
  qualifier: lambda.version,
  provisionedConcurrentExecutions: 2,
});
```

The snippet sets the provisioned concurrency for `mylambda` to a fixed value of `2`. Note that concurrency is provisioned per Lambda version, and it can't be set for the `$LATEST` version alias. Therefore, I set the `publish` attribute of my Lambda to `true` and reference a specific version of the Lambda in provisioning.

## Dynamic Provisioned Concurrency

> A fixed provisioned concurrency works good for a stable workloads.

Reviewer: ```suggestion
A fixed level of provisioned concurrency works well for stable workloads.
```

{{< figure src="./steady.png" caption="Fixed provisioned concurrency for uniform workloads" >}}

> However, many workloads fluctuate a lot. Extreme elasticity and lack of configuration parameters have always been the essential benefits of AWS Lambda. It works great if you can tolerate the cold starts that come during scale-out. If not, you can explore more advanced scenarios of provisioning concurrency.

Reviewer: ```suggestion
However, many workloads fluctuate a lot. Extreme elasticity and lack of configuration parameters have always been the essential benefits of AWS Lambda. It works great if you can tolerate the cold starts that come during scale-out. If not, you can explore more advanced scenarios of provisioning concurrency dynamically.
```

Instead of choosing a permanently fixed value, you can configure provisioned concurrency to autoscale. The first required component is the autoscaling target:

```ts
const resourceId = pulumi.interpolate`function:${lambda.name}:${lambda.version}`;
const target = new aws.appautoscaling.Target("target", {
    resourceId,
    serviceNamespace: "lambda",
    scalableDimension: "lambda:function:ProvisionedConcurrency",
    minCapacity: 1,
    maxCapacity: 10,
});
```

> The second component is the autoscaling policy, which defines the conditions when scaling starts. There are two important way

Reviewer: Incomplete sentence and/or section?

Reviewer 2: Indeed... restoring as well, and checking the rest of the document :(

Author: I'm not sure what happened here. A lot of the edits I made are no longer there.

    resourceId,
    serviceNamespace: "lambda",
    scalableDimension: "lambda:function:ProvisionedConcurrency",
    policyType: "TargetTrackingScaling",
    targetTrackingScalingPolicyConfiguration: {
        targetValue: 0.8,
        predefinedMetricSpecification: {
            predefinedMetricType: "LambdaProvisionedConcurrencyUtilization",
        },
    },
}, { dependsOn: [target]});
```

Currently, there are issues with autoscaling based on the metrics, which makes provisioned concurrency scale less effectively than what the chart above shows. Hopefully, Amazon will fix these issues soon.

## Pricing

While hand-crafted Lambda warmers are virtually free, provisioned concurrency can be costly. The new pricing is an integral part of the change: Instead of purely per-call model, AWS charges per hour for provisioned capacity.

You would pay $0.015 per hour per GB of provisioned worker memory, even if a worker handled zero requests.
The per-invocation price gets a discount: $0.035 per GB-hour instead of the regular $0.06 per GB-hour. This change means that fully-utilized workers would be cheaper if provisioned compared to on-demand workers.

> 

Reviewer: Not necessary - but I wonder if there's a nice graphic that could be included here that shows the two price curves based on % of time the worker is utilized?

Reviewer 2: Should be relatively easy to make if we don't publish for the next 3-4 hours

Reviewer 2: Can only do this tomorrow

Reviewer 2: @lukehoban @spara Added a chart and a paragraph below, please review if it's clear enough.

---
title: "Provisioned Concurrency: Insurance Against Cold Starts in AWS Lambda"
date: 2019-12-18
meta_desc: "Using Pulumi to configure AWS Lambda provisioned concurrency"
meta_image: meta.png
authors: ["mikhail-shilkov"]
tags: ["AWS", "Serverless"]
---

> AWS has recently [announced](https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/) the launch of **Provisioned Concurrency**, a new feature of AWS Lambda that intends to solve the problem of cold starts. In this article, we take another look at the problem of latency-critical serverless applications, and how the new feature impacts the status-quo.

Author: ```suggestion
AWS recently [announced](https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/) the launch of **Provisioned Concurrency**, a new feature of AWS Lambda that intends to solve the problem of cold starts. In this article, we take another look at the problem of latency-critical serverless applications, and how the new feature impacts the status-quo.
```

## Provisioned Concurrency

At re:Invent 2019, AWS introduced Lambda **Provisioned Concurrency**---a feature to work around cold starts. The base concurrency model doesn't change. However, you can request a given number of workers to be always-warm and dedicated to a specific Lambda.

Here is an example of configuring the provisioned concurrency with Pulumi in TypeScript:

```ts
const lambda = new aws.lambda.Function("mylambda", {
    code: new pulumi.asset.FileArchive("./app"),
    handler: "handler",
    role: role.arn,
    publish: true,
});

const fixed = new aws.lambda.ProvisionedConcurrencyConfig("fixed-concurrency", {
  functionName: lambda.name,
  qualifier: lambda.version,
  provisionedConcurrentExecutions: 2,
});
```

The snippet sets the provisioned concurrency for `mylambda` to a fixed value of `2`. Note that concurrency is provisioned per Lambda version, and it can't be set for the `$LATEST` version alias. Therefore, I set the `publish` attribute of my Lambda to `true` and reference a specific version of the Lambda in provisioning.

## Dynamic Provisioned Concurrency

> A fixed provisioned concurrency works good for a stable workloads.

Author: ```suggestion
A fixed provisioned concurrency works well for stable workloads.
```

        targetValue: 0.8,
        predefinedMetricSpecification: {
            predefinedMetricType: "LambdaProvisionedConcurrencyUtilization",
        },
    },
}, { dependsOn: [target]});
```

Currently, there are issues with autoscaling based on the metrics, which makes provisioned concurrency scale less effectively than what the chart above shows. Hopefully, Amazon will fix these issues soon.

## Pricing

While hand-crafted Lambda warmers are virtually free, provisioned concurrency can be costly. The new pricing is an integral part of the change: Instead of purely per-call model, AWS charges per hour for provisioned capacity.

You would pay $0.015 per hour per GB of provisioned worker memory, even if a worker handled zero requests.

The per-invocation price gets a discount: $0.035 per GB-hour instead of the regular $0.06 per GB-hour. This change means that fully-utilized workers would be cheaper if provisioned compared to on-demand workers.

Be careful to clean up your resources after any experiments: Leaving running workers with  provisioned concurrency can be expensive! Using `pulumi destroy` removes resources after you finish experimenting.

## Conclusion

> Provisioned concurrency brings long-awaited solution to cold starts in AWS Lambda. However, it comes at a cost in terms of a direct impact on billing and management overhead to provision concurrency at optimal levels.

Author: ```suggestion
Provisioned concurrency brings the long-awaited solution to cold starts in AWS Lambda. However, it comes at a cost in terms of a direct impact on billing and management overhead to provision concurrency at optimal levels.
```

> For high-load functions where utilization is continuously above a known level of requests, it makes sense to set the provisioned concurrency to that level in order to save money and have the guarantee of warm workers.

Author: ```suggestion
For high-load functions where utilization is continuously above a known level of requests, it makes sense to set the provisioned concurrency to that level to save money and have the guarantee of warm workers.
```

> If your Lambda hosts a latency-sensitive API, especially with runtimes like Java and .NET, you should strive to find the right balance between the percentage of requests that result in a cold start and the money spent on concurrency. Autoscaling should help once AWS has fixed the initial glitches that slipped into their current services, are fixed.

Author: ```suggestion
If your Lambda hosts a latency-sensitive API, especially with runtimes like Java and .NET, you should strive to find the right balance between the percentage of requests that result in a cold start and the money spent on concurrency. Autoscaling should help once AWS has fixed the initial glitches that slipped into their current services.
```
